lib/client/escape/reader.go

```diff
@@ -0,0 +1,187 @@
+	out          io.Writer
+	onDisconnect func(error)
+	bufferLimit  int
+
+	// cond protects buf and err and also announces to blocked readers that
+	// more data is available.
+	cond sync.Cond
+	buf  []byte
+	err  error
+}
+
+// NewReader creates a new Reader to catch escape sequences from 'in'.
+//
+// Two sequences are supported:
+// - "~?": prints help text to 'out' listing supported sequences
+// - "~.": disconnect stops any future reads from in; after this sequence,
+//   callers can still read any unread data up to this sequence from Reader but
+//   all future Read calls will return ErrDisconnect; onDisconnect will also be
+//   called with ErrDisconnect immediately
+//
+// NewReader starts consuming 'in' immediately in the background. This allows
+// Reader to detect sequences without the caller actively calling Read (such as
+// when it's stuck writing out the received data).
+//
+// Unread data is accumulated in an internal buffer. If this buffer grows to a
+// limit (currently 10MB), Reader will stop permanently. onDisconnect will get
+// called with ErrTooMuchBufferedData. Read can still be called to consume the
+// internal buffer but all future reads after that will return
+// ErrTooMuchBufferedData.
+//
+// If the internal buffer is empty, calls to Read will block until some data is
+// available or an error occurs.
+func NewReader(in io.Reader, out io.Writer, onDisconnect func(error)) *Reader {
+	r := newUnstartedReader(in, out, onDisconnect)
+	go r.runReads()
+	return r
+}
+
+// newUnstartedReader allows unit tests to mutate Reader before runReads
+// starts.
+func newUnstartedReader(in io.Reader, out io.Writer, onDisconnect func(error)) *Reader {
+	return &Reader{
+		inner:        in,
+		out:          out,
+		onDisconnect: onDisconnect,
+		bufferLimit:  readerBufferLimit,
+		cond:         sync.Cond{L: &sync.Mutex{}},
+		// note: no need to pre-allocate buf, it will allocate and grow as
+		// needed in runReads via append.
+	}
+}
+
+func (r *Reader) runReads() {
+	// prev contains up to 2 characters from previous Read to catch sequences
+	// spanning multiple reads.
+	prev := []byte{'\r'}
+	// Use a small read buffer. We're reading terminal input, so there will
+	// only be a few characters at a time.
+	readBuf := make([]byte, 128)
+	for {
+		n, err := r.inner.Read(readBuf)
+		if err != nil {
+			r.setErr(err)
+			return
+		}
+
+		// Check for escape sequences.
+		keys := append(prev, readBuf[:n]...)
+		switch {
+		case bytes.Contains(keys, []byte(escapeHelp)):
+			r.printHelp()
+		case bytes.Contains(keys, []byte(escapeDisconnect)):
+			r.setErr(ErrDisconnect)
+			return
+		}
+
+		// Reset prev to the last 2 characters read.
+		prev = keys
+		if len(prev) > 2 {
+			prev = prev[len(prev)-2:]
+		}
+
+		// Add new data to internal buffer.
+		r.cond.L.Lock()
+		if len(r.buf)+n > r.bufferLimit {
+			// Unlock because setErr will want to lock too.
+			r.cond.L.Unlock()
+			r.setErr(ErrTooMuchBufferedData)
+			return
+		}
```

Comment:
Seems preferable to propagate backpressure, rather than greedily consume bytes and then error on an arbitrary limit.  This custom logic could probably be replaced with an `io.Pipe` which would handle backpressure naturally.

Reply:
That makes sense for the usual connection plumbing.
But our source in this case is a human at the keyboard. We want to react to their disconnect sequences as soon as possible, even if some data is blocked on being written to the remote side.

Let me know if you think my reasoning is wrong, I may be misunderstanding some things!

Reply:
You're absolutely right, and the top-level comments explain your reasoning for making this type greedy!  I took this snippet out of context.  Carry on.